src/System.Management.Automation/engine/lang/codegen.cs

```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        /// handling escaping of wildcard patterns when argument is not already taken literally.
+        /// For example: QuoteArgument(userContent, quoteInUse, isLiteralArgument)
+        /// </summary>
+        /// <param name="value">The content to be used as an argument value taken literally.</param>
+        /// <param name="quoteInUse">The character to be quoted with.</param>
+        /// <param name="isLiteralArgument">Treat the argument as taken literally, wildcard escaping not required.</param>
+        /// <returns>Content quoted and escaped if required for use as an argument value.</returns>
+        public static string QuoteArgument(string value, char quoteInUse, bool isLiteralArgument)
+        {
+            return string.IsNullOrEmpty(value) ?
+                string.Empty :
+                QuoteArgument(isLiteralArgument ? value : WildcardPattern.Escape(value), quoteInUse);
+        }
+
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly
+        /// For example: QuoteArgument(userContent, quoteInUse)
+        /// </summary>
+        /// <param name="value">The content to be used as an argument value taken literally.</param>
+        /// <param name="quoteInUse">The character to be quoted with.</param>
+        /// <returns>Content quoted and escaped if required for use as an argument value.</returns>
+        public static string QuoteArgument(string value, char quoteInUse)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
+
+            char quoteToUse = quoteInUse;
+            if (quoteToUse == 0)
+                if (ShouldArgumentNotBeBareword(value))
+                    // argument value not compatible with bareword
+                    quoteToUse = '\'';
+                else
+                    // return unmodified argument value
+                    return value;
+
+            // quote and escape argument as needed
+            return quoteToUse + (quoteToUse.IsDoubleQuote() ? EscapeDoubleQuotedStringContent(value) :
+                EscapeSingleQuotedStringContent(value)) + quoteToUse;
+        }
+
+        // brute force method to determine when an argument's value cannot be used bareword
+        private static bool ShouldArgumentNotBeBareword(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                return false;
+
+            // Rules for bareword arguments:
+            //   characters that cannot be in the first position: '@','#','<','>'
+            //   Pattern that cannot be in the first position: /[1-6]>/
+            bool requiresQuote = "@#<>".Contains(value[0]) ||
+                value.Length > 1 && '1' <= value[0] && value[0] <= '6' && value[1] == '>';
+
+            if (!requiresQuote)
+            {
+                bool lastCharWasDollar = false;
+                foreach (char c in value)
```

Comment:
Please use `for` cycle.

Reply:
Explain `for cycle`?

I built the loops referencing the example in the same file, `EscapeSingleQuotedStringContent()`

Reply:
No need to use an enumarator to access chars from a string. 
I means value[i].

Reply:
so a for loop of

```C#
// skipping first char (special case)
for( int i = 1; i < value.Length; i++ )
{
    value[i]  ...
}
```

will be faster than the enumerator?  Is this something that can only be performance tested in release?

Reply:
> will be faster than the enumerator? 

You could test with BenchmarkDotNet. 
I say only that we have no need to allocate an enumerator in the simple code.

Reply:
I finally got to do some benchmarks.  It does appear that the method 'indexing' is probably both the fastest and the most flexible, particularly when the strings are short.  Once the strings reach 10000's of characters, 'indexing' and 'foreach' methods differences become negligible, though system background tasks had a larger effect at skewing the results between runs.